New RFC: 37/WAKU2-NOISE-SESSIONS (#504)
vacp2p/rfc: content/docs/rfcs/37/README.md

This post provides an overview on how we can possibly implement and manage one or multiple Noise sessions in Waku.

# Preliminaries

We assume that two users, e.g. Alice and Bob, successfully completed a Noise handshake.

Using [Noise terminology]((http://www.noiseprotocol.org/noise.html)), at the end of the handshake they will share:
- two _Cipher States_ `CSOutbound` and `CSInbound`, to encrypt and decrypt outbound and inbound messages, respectively;
- a handshake hash value `h`.

As suggested in Noise specifications in regards to [Channel Binding](http://www.noiseprotocol.org/noise.html#channel-binding), 
we can identify a Noise session with a `session-id` derived from the handshake hash value `h` shared on completion of a Noise handshake.

More specifically, when Alice and Bob call [Split()](http://www.noiseprotocol.org/noise.html#the-symmetricstate-object) in order to derive the two final encryption and decryption Cipher States, 
they further compute `session-id  = HKDF(h)` using the supported key derivation function `HKDF`.

Such `session-id` will uniquely identify the Noise cryptographic session instantiated on completion of a Noise handshake,

> which would then consists of the tuple `(session-id, CSOutbound, CSInbound)`. 

Reviewer: ```suggestion
which would then consist of the tuple `(session-id, CSOutbound, CSInbound)`. 
```

Once a Noise session is instantiated, 

> any further encrypted message between Alice and Bob within this session is exchanged on a `contentTopic` with name `/wakunoise/1/<ct-id>/proto`, 

Reviewer: Wondering what proto is. The name of an encapsulated protocol?
The noise protocol ID?

Author: The encoding! This follows https://rfc.vac.dev/spec/23/#pubsub-topic-format

> where `ct-id = Hash(Hash(session-id))`.

Reviewer 2: Any reason for double hash?

Author: Sure, it is to avoid an attacker to claim that he was a party in a previous session while allowing _public stale_ of sessions. To publicly stale a certain session with content topic `ct-id = Hash(Hash(session-id))`, you reveal `Hash(session-id)`, a pre-image, whose validity can publicly verified. When instantiating a new session, if you want to do channel binding with a previous one (either stale or active), you use `session-id`, which is known only to parties either if the previous session is active of was declared stale.

Reviewer 2: >  it is to avoid an attacker to claim that he was a party in a previous session

Can you please expand on it? In which scenario would it be considered as attack? (we can also discuss in our 1:1)

Author: This relates to channel binding and reactivation of Noise session. The session-id allows re-activation, hence an attacker would be able to show that he was involved in a previous handshake while binding a new channel if we only consider Hash(session-id) and stale with session-id.

# Session states

A Noise session corresponding to a certain `session-id`:
- is always **active** as long as it is not marked as **stale**. 
For an active `session-id`, new messages are published on the content topic `/wakunoise/1/<ct-id>/proto`;  

> - is marked as **stale** if a [session termination message](https://rfc.vac.dev/spec/35/#session-termination-message) containing `Hash(session-id)` is published on the content topic `/wakunoise/1/<ct-id>/proto`. 

Reviewer 3: How would this interact with protocols for resource-restricted nodes, e.g. `filter`, where clients subscribe to only receive messages on content topics related to applications they're interested in? Would all clients of applications using Waku Noise sessions have to subscribe to `/wakunoise/1` for key exchange? Would there be any application-specific sessions over the application content topic?

Reviewer: Would it make sense to perform the handshake outside of relay and such via a separate 1:1 protocol?

Author: @jm-clius yes! Having application specific content topics and also ad-hoc content topics to differentiate between different stages of the handshakes (e.g. handshake-start/key-bundle retrieval, finalization/communication, etc.) would improve a lot efficiency, especially for limited resource devices. Thanks for the suggestion!

Author: @kaiserd What do you mean exactly? I think some security guarantees might change with 1:1 protocols (e.g. censorship resistance, connection reliability) and would probably require the two users to be online at the same time. But I might have got wrong what you mean

Author: @jm-clius tried to address this in https://github.com/vacp2p/rfc/commit/5fc0fcdcf76bd076d289568706bc277a05ba5f16

> - is marked as **stale** if a [session termination message](https://rfc.vac.dev/spec/35/#session-termination-message) containing `Hash(session-id)` is published on the content topic `/wakunoise/1/<ct-id>/proto`. 

Reviewer 3: How would this interact with protocols for resource-restricted nodes, e.g. `filter`, where clients subscribe to only receive messages on content topics related to applications they're interested in? Would all clients of applications using Waku Noise sessions have to subscribe to `/wakunoise/1` for key exchange and then continue encrypted conversation over the application content topic? Could make sense, but want to be sure I see the full picture here.

Author: See https://github.com/vacp2p/rfc/pull/504#discussion_r880283108

Session information relative to stale sessions  MAY be deleted from users' device, unless required for later channel binding purposes.

When a Noise session is marked as stale, it means that one party requested its termination while being online, 
since publication of a hash pre-image for `ct-id` is required (i.e. `Hash(session-id)`).

> Currently, it is not possible to mark a Noise session as stale when `session-id` is lost or gets corrupted in users' devices.

Reviewer 2: It is also not possible to communicate if the session-id or its H(H(session-id)) is lost, so worth emphasizing that the session-id and its hashes MUST be persisted properly.

Author: Actually this sentence emphasize something different. The Noise session information, consisting of the `session-id` and the two Cipher states, is clearly required not only to agree on a place where parties communicate (the content topic) but to encrypt/decrypt messages in the first place: so persistence of `session-id` is as important as encryption keys when communicating.

Here, instead, we refer to the possibility to stale a session: if users lose session information (session-id + Cipher state) the session will remain active, at least until the other party takes some actions. But only `sessiond-id`is required (or, in principle `Hash(session-id)`) to publicly stale a session, rather than the full Noise session information.

Reviewer 2: Thanks for the explanation, but my point was exactly the same, you have mentioned one but not the other i.e., the session id should be persisted
- To be able to communicate with the other party
- To be able to properly exit the communication with the other party
You have emphasized the second point, but not the first one.

Author: Not only the session-id, but the whole Noise session, that is session-id + Cipher states. I guess it is assumed that Noise session has to be persisted, since contains all the cryptographic information to either retrieve and process messages. This may be made explicit maybe earlier, but at this point, instead, the discussion is about staling sessions. Do you agree?

Author: Addressed in https://github.com/vacp2p/rfc/commit/f4c6b40ed3647120770a6330355c09e854d45e0c

Reviewer 2: > This may be made explicit maybe earlier, but at this point, instead, the discussion is about staling sessions. Do you agree?

Sure, not right under this section, any place in the RFC that makes the most sense!

However, since `session-id` is shared between Alice and Bob, 
one party MAY decide to mark a Noise session as stale if no message from the other end was received within a certain fixed time window. 

The above mechanism allows a Noise session to be marked as stale either privately or publicly, 
depending if `Hash(session-id)` is sent on `/wakunoise/1/<ct-id>/proto` to the other party in encrypted form or not, respectively. 

When a Noise session is publicly marked as stale, 

> network peers MAY discard all [stored](https://rfc.vac.dev/spec/13/) messages addressed to the content topic `/wakunoise/1/<ct-id>/proto`. 

Reviewer 2: Can you please clarify what _network peers_ mean exactly?

Reviewer 2: I think it would complicate things by asking store nodes to watch for the content topic of the waku messages. The content topic of a waku message should not have any specific meaning for the nodes running the store protocol otherwise they need to stay updated by the sematic of the content topic of all the applications. This only tangles the two protocols.

Author: network peers here refers mainly to store nodes (but I kept their definition generic, since the are not capable yet of that, as you described). We had internal discussion with @jm-clius, that in the [near future](https://github.com/status-im/nwaku/issues/937) store nodes may become able to see the content topic of a Waku message. Hence the resources optimization given by staling sessions was designed having this possibility in mind, by considering also that this is a RAW initial proposal. 

In any case, at the moment discard is optional, to both comply to current implementation and eventually address the fact that discarding messages creates other undesirable problems.

Reviewer 3: To clarify, the feature was not to allow `store` nodes to "see" content topics. It merely allows a `store` to be selective in terms of which content topics it stores for _before storing anything._ In other words, any logic that would make a `store` discard already-stored messages on certain content topics _after_ the fact would still have to be implemented. This is possible, but may indeed over-complicate things.

Reviewer 2: > When a Noise session is publicly marked as stale, 
network peers MAY discard all [stored](https://rfc.vac.dev/spec/13/) messages addressed to the content topic `/wakunoise/1/<ct-id>/proto`.

I do not see this as a viable solution,  imagine a large number of applications running on top of waku and each coming up with a different semantic for the content topic namings, it is not reasonable to expect that store nodes to be aware of such external semantics and update their database/filter their incoming messages accordingly.  

> this is a RAW initial proposal.

I would suggest not including it in the RAW version

Author: Does this approach (already committed) help in this case? https://github.com/vacp2p/rfc/pull/504#discussion_r880283108

In this the case and in order for parties to retrieve any eventually delayed message, 
peers SHOULD wait a fixed amount of time before discarding stored messages corresponding to a stale Noise session. 

A stale Noise session cannot be directly marked as active
and parties are required to instantiate a new Noise session if they wish to communicate again. 

> However, parties can optionally include the `session-id` corresponding to a stale Noise session in the [prologue information](https://noiseprotocol.org/noise.html#prologue) employed in the Noise handshake they execute to instantiate their new Noise session. 

Reviewer 2: > [prologue information](https://noiseprotocol.org/noise.html#prologue)

can you please supply a link to our own spec?

Reviewer 2: The term _reactivating the noise session_ is a bit confusing. By using the Noise prologue input, the `session id` is included in the `h` variable of the handshake, and then the parties start the _handshake_ by such `h` value, so how it is relevant to reactivating a _Noise session_?

Author: The term "re-activate" is used in this sentence

> This effectively emulates a mechanism to "re-activate" a stale Noise session by binding it to a newly created active Noise session.

and I used double quotes on purpose to stress that such term is only to provide an intuition of what is happening in the technical description that follows.

Here the word "re-activate" made sense to me because I was talking about the scenario where a session remains active for a long period of time but with no message received from the other party (think of a device that remains offline 1 year or got stolen). Then the other party can unilaterally decide to mark such session as stale (one reason I was considering this possibility here, is that in the future store nodes may provide as a service storage of messages for longer period of time than current 30 days). If, by chance, the offline device comes back online later on, and wants to communicate with the other party again (which in the meanwhile marked the session stale), he then need to instantiate a new session with him, but he can do it in a way that the new session is bind to the old staled one, in fact emulating a session _re-activation_.

Reviewer 2: Thanks for the explanation, what is the benefit such re-activating from your PoV? Why not establishing a new one? 
BTW, the session-id must be also persisted by the other party in order to be able to engage in the reactivated session, right?

Author: The advantage is to bind new messages to the any previous executed handshake. Imagine if you keep and update the handshake hash for any exchanged message, such binding would be stronger. Of course it has to be persisted in order to be reactivated: clarified it in https://github.com/vacp2p/rfc/pull/504/commits/9cff4db8f1ade87e9dada9cc7eb93f6cce55b13f

This effectively emulates a mechanism to _"re-activate"_ a stale Noise session by binding it to a newly created active Noise session.

In order to reduce users' metadata leakage, it is desirable (as suggested in [35/WAKU2-NOISE](https://rfc.vac.dev/spec/35/#after-handshake)) that content topics used for communications change every time a new message is exchanged.
This can be easily realized by employing a key derivation function to compute a new `session-id` from the previously employed one (e.g. `session-id = HKDF(prev-session-id)`),
while keeping the Inbound/outbound Cipher States, the content topic derivation mechanism and the stale mechanism the same as above.

> In this case, when one party sends **and** receives at least one message, 

Reviewer 2: ```suggestion
In this case, when one party sends **or** receives at least one message, 
```

Author: No, the **and** is correct. 

If, for example, a user sends a message and marks as stale all previous sessions, after a period of time store nodes may discard all messages corresponding to these stale sessions. If one party was temporary offline while the other party was sending many messages, with the exception of the last message, the others may not be delivered due to marking previous sessions stale. Once online again, he will then not be able to receive/decrypt messages due to these missed messages.

If, instead, a message is received after a message is sent, or vice-versa, it means that at that point the two party have synced session information, which allows them to encrypt/decrypt and move to new content topics. Hence, all session relative to messages exchanged before these last two can be marked as stale, since the session information is proved to be synced at least up to this point.

> he SHALL publicly mark as stale all Noise sessions relative to messages exchanged before the earlier of these two send/receive events.

Reviewer 2: why does it need to be marked as stale? it would make sense in the one content topic per session, but in this case, the two parties are already aware that after each message the prior content topic/ session-id is stale

Author: Parties are aware, but not store nodes which, in order to save resources, may discard messages corresponding to stale sessions. See https://github.com/vacp2p/rfc/pull/504#discussion_r878676315.

Reviewer 2: Please see https://github.com/vacp2p/rfc/pull/504#discussion_r880957935

# Multi-Device support

Alice and Bob might possess one or more devices (e.g. laptops, smartphones, etc.) they wish to use to communicate.

> In the following we assume Alice and Bob to possess $N$ and $M$ devices, respectively.

Reviewer 2: ```suggestion
In the following, we assume Alice and Bob to possess $N$ and $M$ devices, respectively.
```

Author: Added comma in https://github.com/vacp2p/rfc/pull/504/commits/2cb76227edc8bfd061999a7ae42c89fecbb2b0a5

Since a Noise session contains cryptographic material required to encrypt and decrypt messages exchanged on a pre-defined content topic derived from a `session-id`, 
messages should be encrypted and decrypted within the Noise session instantiated between the currently-in-use sender's and receiver's device.

> This is achieved through two main supported session management mechanisms that we called `N11M` and `NM`, respectively.

Reviewer 2: What do these names `N11M` and `NM` stand for? are they abbreviations of something else?

Author: I agree that the names are not the best possible. These names sketch the Noise session graph layout: 
- In NM all Alice's N devices have a session with all Bob's M devices, that is: each Alice's N devices <-> each Bob's M devices.
- In N11M, there is a 1-to-1 session shared among all N and all M Alice's and Bob's devices, that is: (Alice N devices) <-> (Bob M devices). 

I thought that pictures would help understanding the reason for such names, but by being just names I didn't want to go into the details of their meaning.

Reviewer: I actually like the names. They are kind of in line with noise protocol names.

Reviewer 2: > In NM all Alice's N devices have a session with all Bob's M devices, that is: each Alice's N devices <-> each Bob's M devices.
In N11M, there is a 1-to-1 session shared among all N and all M Alice's and Bob's devices, that is: (Alice N devices) <-> (Bob M devices).

Thanks for the clarification!

## The $N11M$ session management mechanism

In a $N11M$ setting, each party's device shares the same Noise session information used to encrypt and decrypt messages exchanged with the other party.

![](./N11M.png)

More precisely, once the first Noise session between any of Alice's and Bob's device is instantiated, 

> its session information is securely propagated to all other devices, 

Reviewer 2: Are other devices of the same person disconnected from the network? is only one device the entry point of other devices?

Author: Once Noise sessions are instantiated between devices, they can go all offline, since sync information would be propagated on the agreed content topic and made available by store nodes. It is one device at a time that communicates with the other party. If multiple devices (with same session information) send a message simultaneously, then the first received message will be the only decrypted correctly, while the others will have an invalid nonce for the receiver. Hence, based on the session information propagated and how Noise encryption works, only one device per message can communicate to the other party.

Reviewer 2: > one device per message can communicate to the other party.

This is not clear from the description, please clarify it.

Author: Addressed in https://github.com/vacp2p/rfc/commit/f4c6b40ed3647120770a6330355c09e854d45e0c

which then become able to send and receive new messages on the content topic associated to such session.

The most updated session information between Alice and Bob is propagated in encrypted form to other devices, 

> using previously instantiated Noise sessions. 

Reviewer 2: Do you mean `previously instantiated` or the `current`?

Reviewer 2: Please clarify this paragraph, is is somewhat ambiguous

Author: I meant that when Alice instantiates a session with Bob or updates such information by simply exchanging messages (e.g. new content topic, nonces, rekeys, etc.), all Alice's devices (resp. Bob's devices) that wants to receive such updated session information, are required to have an already instantiated Noise session between them in order to receive it in encrypted form. What do you find ambiguous, exactly?

Reviewer 2: > all Alice's devices (resp. Bob's devices) that wants to receive such updated session information, are required to have an already instantiated Noise session between them in order to receive it in encrypted form.

Thanks for the clarification, the part I quoted above is missing in the text.

Author: Addressed in https://github.com/vacp2p/rfc/commit/f4c6b40ed3647120770a6330355c09e854d45e0c

The propagated session information corresponds to the latest session information stored on the device currently communicating with (any of the devices of) the other party.

We note that sessions information is propagated only among devices belonging to the same party and not with other party's devices. 
Hence, Alice has no knowledge on the number of devices Bob is using and vice versa. 

When any device marks a Noise session between Alice and Bob as stale, 

> all (updated) devices will implicitly mark such session as stale.

Reviewer 2: ```suggestion
all (updated) devices will mark such a session as stale upon synchronization with the active device.
```

Author: The stale is implicit here, which means that the session is marked stale with no previous/further action from other devices. Since the device that marked the session stale does it by publishing a message on the corresponding content topic, and all Alice's (updated) device will share the same session information and will read such message, devices will see that the session was marked stale (even if they didn't to directly) and no further action is required from them (besides application specific actions such as deleting session information etc.).

Reviewer 2: The ambiguity is around the `implicit` term, Do you mean they `internally`/`'locally` mark it as stale?

Author: I see! The fact is that other devices don't mark it as stale. The marking is done online by one device, will others will just see the session to be marked as stale. So, since the session information is shared, this is equivalent to the fact that each device marked it as stale, but only one actually did it: all the others did it _implicitly_! I cannot think about another word to express this!

Reviewer 2: I see your point, maybe `they consider the corresponding Noise session as stale and act accordingly`, though, no pressure!

Author: Addressed in https://github.com/vacp2p/rfc/pull/504/commits/9cff4db8f1ade87e9dada9cc7eb93f6cce55b13f

> In case a Noise session between two devices belonging to the same party is marked as stale, 

Reviewer 2: Have you explained that devices of the same person instantiate a Noise session?

Author: Yes, it is few lines earlier: https://github.com/vacp2p/rfc/blob/2cb76227edc8bfd061999a7ae42c89fecbb2b0a5/content/docs/rfcs/37/README.md?plain=1#L97-L98

Reviewer 2: >  instantiated Noise sessions.  

Can you please explain when does/did the `previously Noise sessions` happen? 
Update: Please see this comment https://github.com/vacp2p/rfc/pull/504#discussion_r880971264

Author: Addressed in https://github.com/vacp2p/rfc/commit/f4c6b40ed3647120770a6330355c09e854d45e0c

> such two devices stop to reciprocally propagate any information regarding Noise sessions instantiated with other parties.

Reviewer 2: What is the assumption about the handshake phase between the two parties' devices? let's say 2 devices of Alice simultaneously enter a handshake with one or more devices of Bob? they all arrive at the same content topic and they will collide.

By the way, it would be interesting and useful to look into scenarios where devices of the same person can be online at the same time and not in sync with each other.  How devices of the same person are differentiated?

Author: > What is the assumption about the handshake phase between the two parties' devices? let's say 2 devices of Alice simultaneously enter a handshake with one or more devices of Bob? they all arrive at the same content topic and they will collide.

Collisions are practically impossible as long as the handshake involves DH operations over ephemeral keys (not considering that each device may have a different static key too). If two devices simultaneously instantiate a session with the other party, these two party-to-party sessions will have two distinct session-id and Cipher states pairs, in turn shared among all parties' devices. Having two concurrent session with one party is a requirement for some applications, e.g. being able to download multiple files from the same source.

> By the way, it would be interesting and useful to look into scenarios where devices of the same person can be online at the same time and not in sync with each other. How devices of the same person are differentiated?

There is no need to. Simultaneous messages from different devices will be automatically ordered thanks to increasing Cipher States nonces, that is if are encrypted with same key and same nonce only the first one received will be correctly decrypted (you may implement an ACK mechanism on top, but this is left to the application layer). See https://github.com/vacp2p/rfc/pull/504#discussion_r878679430 for a related comment. If a device session information is not synced with the other party's session, received messages will simply be discarded as long as this device does not get the most updated session information shared by other devices.

Reviewer 2: > Collisions 

Here is what I meant by collision:
Alice has two devices A1, and A2, where A1 and A2 are not synchronized
Bob has one device B1
A1 does the handshake with B1 and enters a Noise session with B1
A2 also does a handshake with B1 and enters a Noise session  with B1
Now, A1 and A2 are concurrently talking with B1 in two separate sessions
From B1's point of view, it is talking with Alice on two different Noise sessions. Now, assume this is a chat app, it is like Bob is talking to two Alices. It might make sense for Bob i.e., B1 to check whether it already has a connection with Alice or not and e.g., inform A2 of such communication OR refuses to instantiate a connection with A2 if there is another active session. Although, this can be an application concern. Please let me know if that makes sense. 

BTW,  I meant to consider such scenarios in the future (not in this PR).

Author: Sure, this is possible and we want it to be possible (indeed, is an application concern). The two can easily identified, for example, by sending after handshake a random string derived by uniquely identifying all Alice and Bob devices (e.g. an extra static key not employed by authentication purposes). Sure, when we have clearer application scenario for session management we can address this better by future PR.

This can be mitigated by adopting techniques similar to the the ones proposed in [35/WAKU2-NOISE](https://rfc.vac.dev/spec/35/#after-handshake), 
where encryption keys are changed every time a new message is exchanged.

This session management mechanism is loosely based on the paper ["Multi-Device for Signal"](https://eprint.iacr.org/2019/1363.pdf).

## The $NM$ session management mechanism

In a $NM$ setting, we require all of $N$ Alice's devices to have an active Noise session with each of Bob's $M$ devices, 
for a total of $NM$ concurrently active Noise sessions between Alice and Bob.

![](./NM.png)

A message is sent from the currently-in-use sender's device to all recipent's devices, 
by properly encrypting and sending it to the content topics of each corresponding active Noise session. 

We note that this allows the recipient to receive a message on all his devices simultaneously. 
However, on the sender side, only the device which effectively sent the message will know its full content.

If it is required for sent messages to be available on all sender's devices, 
each pair of sender's devices SHOULD have an active Noise session used for syncing purposes: 

> this sums up to a total of $N-1$ and $M-1$ extra Noise sessions instantiated on each Alice's and Bob's device, respectively.

Reviewer 2: We cannot guarantee which device is active now, hence I'd say there should be a session for each pair of devices of the same person, like a full graph, thus N*(N-1)/2 sessions.

Author: Indeed, this is what is written: $N-1$ extra sessions per device, that is every device has a session with all the remaining $N-1$ devices, for a total of $N(N-1)/2$ unique sessions.
